cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/LinkAnalysis.scala

```diff
@@ -0,0 +1,97 @@
+ * our algorithms.  We will not be storing the results every step of the way--instead we find the iteration
+ * that first satisfies the condition of exceeding maximum iterations or falls below the error threshold.
+ * @param params The set of all parameters passed into our algorithm
+ * @param modelName This must be provided by all subclasses for logging purposes.  This should not be
+ *                  left up to the user to define.  This must be provided by the implementer.
+ * @tparam T `LinkAnalysis` must be generically typed by `Iteration` or one of its subclasses.  An `Iteration`
+ *           holds all of the pertinent information for a given algorithm.
+ */
+abstract class AbstractLinkAnalysis[T <: IterationState](graph: DirectedGraph[Node],
+    params: Params, modelName: String) {
+
+  protected val log = Logger.get(modelName)
+
+  protected val maxIterations = params.maxIterations
+  protected val tolerance     = params.tolerance
+
+  /**
+   * Run a single iteration through our algorithm.
+   * @param iteration The starting iteration that our algorithm will be applied to.
+   * @return A new iteration.
+   */
+  def iterate(iteration: T): T
+
+  /**
+   * Provides default initial start values for our algorithms.
+   * @return An default starting iteration.
+   */
+  protected def defaultInitialIteration: T
+
+  /**
+   * Calculate the error between two arrays using either the T1 error or the T2 error.  This is a
+   * convenience method.
+   * @param t1 Flag `true` to calculate the T1 error (the sum of absolute differences between two arrays).  Flag
+   *           `false` to calculate the T2 error (the sum of the squared differences between two arrays).
+   */
+  protected def deltaOfArrays(a: Array[Double], b: Array[Double], t1: Boolean = true): Double = {
+    val rawError = (a zip b).map { case (v1,v2) => if(t1) Math.abs(v1 - v2) else Math.pow(v1 - v2, 2.0) }.sum
+    if (t1) rawError else Math.sqrt(rawError)
+  }
+
+  /**
+   * Run the algorithm to completion according to the parameters passed on instantiation.
+   * @param init The starting point of the iteration.  If no iteration is given, the default start
+   *             is assumed
+   * @return The final iteration.
+   */
+  def run(init: T = defaultInitialIteration): T = {
+    var iters = init.iteration
+    var error = init.error
+    var currentIteration = init
+
+    // Let the user know if they can save memory!
+    if (graph.maxNodeId.toDouble / graph.nodeCount > 1.1 && graph.maxNodeId - graph.nodeCount > 1000000)
+      log.info("Warning - you may be able to reduce the memory usage by renumbering this graph!")
+
+    log.debug(s"Initializing starting ${modelName}...")
+    val progress = Progress(s"${modelName}_init", 65536, Some(graph.nodeCount))
+
+    def terminationCondition: Boolean = if (maxIterations.isDefined)
+      iters < maxIterations.get && error > tolerance
+    else error > tolerance
+
+    while (terminationCondition) {
```

Comment:
minor code change to minimize the number of vars would be to 
1. remove the vars 'iters' and 'error'.
2. s/def terminationCondition/def terminate(currState)/
3. s/while (terminationCondition)/while (!terminate(currentIteration)/